### DISCUSSION TITLE
Dealing with linked data and displaying that in the form.

### DISCUSSION THREAD

**Discussion by Brammm** [react-hook-form/react-hook-form#10256] [Q&A] [answered]
I have a form that gathers Contact details and has the ability to link Contacts to Companies. Below the Contact form fields, I want to display a list of linked companies with some basic Company data (name, vat number...), the ability to unlink them and an autocomplete to look up Companies to link. 

I'm wondering what the best way to go about this is:

Currently, I have something like 
```ts
const schema = z.object({
    first_name: z.string().min(1),
    last_name: z.string().min(1),
    // ... more fields
    companies: z
        .object({
            company_id: z.string(),
        })
        .array(),
});

const {linkedCompanies, ...fields} = dataFromApi;
const defaultValues = {...fields, companies: linkedCompanies.data.map((company) => ({company_id: company.id}))};

const {control} = useForm({
    defaultValues,
    resolver: zodResolver(schema),
});
const {append, fields, remove} = useFieldArray({control, name: 'companies'});

// ... JSX
{fields.map((field, index) => {
    // @ts-ignore
    const company = linkedCompanies.find((company) => company.id === field.company_id);
    if (!company) {
        return null; // Should never happen because fields are constructed from linkedCompanies
    }

    return (
        <div key={field.id}>
            <input key={field.id} type="hidden" {...register(`companies.${index}.id`)} />
            // {company.name} stuff
        </div>
    );
})}
```

Having to find a company entity in another array using data from fields seems... Icky... Is this the best way of doing it? Or should I build a custom Component that holds an array of ids, can add and remove from that array and use a <Controller /> to deal with it?

**Comment by leapful** [comment]
@Brammm 

Based on your description, you might need a list of all companies returned from BE data then just build a Map<ID, Company> on that full list and lookup and render full data for companies that are linked to Contact.

For example:

```
const listCompany = fetch();
const mapCompany = listCompany.reduce((acc, company) => ({ ...acc, [company.id]: company }), {});

{fields.map((field) => {
    const targetCompany = mapCompany[field.company_id];
   
    return ... // render with full data of target company
    
})}
```

**Reply by Brammm** [reply, marked as answer]
Ah yeah, that's definitely better than the find thing I had! I did end up going with the alternate route: made a custom controlled component, combined with `useController` that's passed the full array and an `onChange` prop that gets handled with `field.onChange(companies.map((company) => company.id));`.
